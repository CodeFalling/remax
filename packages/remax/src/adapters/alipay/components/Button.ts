import factory from './factory';

export interface ButtonProps {
  id?: string;
  className?: string;
  size?: 'default' | 'mini';
  type?: 'primary' | 'default' | 'warn';
  plain?: boolean;
  disabled?: boolean;
  loading?: boolean;
  hoverClass?: string;
  hoverStartTime?: number;
  hoverStayTime?: number;
  hoverStopPropagation?: boolean;
  formType?: 'submit' | 'reset';
  openType?: 'share' | 'getAuthorize' | 'contactShare' | 'lifestyle';
  scope?: 'phoneNumber';
  appParameter?: string;
  publicId?: string;
  onTap?: (e: any) => void;
}
export const Button = factory<ButtonProps>('button');

export default Button;
