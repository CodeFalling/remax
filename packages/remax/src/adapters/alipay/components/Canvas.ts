import * as React from 'react';
import factory from './factory';

export interface CanvasProps {
  id?: string;
  style?: React.CSSProperties;
  className?: string;
  width?: string;
  height?: string;
  disableScroll?: boolean;
  onTap?: (e: any) => void;
  onTouchStart?: (e: any) => void;
  onTouchMove?: (e: any) => void;
  onTouchEnd?: (e: any) => void;
  onTouchCancel?: (e: any) => void;
  onLongTap?: (e: any) => void;
}
const Canvas = factory<CanvasProps>('canvas');

export default Canvas;
