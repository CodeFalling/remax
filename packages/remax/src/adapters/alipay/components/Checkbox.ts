import factory from './factory';

export interface CheckboxProps {
  id?: string;
  className?: string;
  value?: string;
  checked?: boolean;
  disabled?: boolean;
  color?: string;
  onChange?: (e: any) => void;
}
const Checkbox = factory<CheckboxProps>('checkbox');

export default Checkbox;
