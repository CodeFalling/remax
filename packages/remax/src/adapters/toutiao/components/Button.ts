import factory from './factory';

export interface ButtonProps {
  size?: 'default' | 'mini';
  type?: 'primary' | 'default';
  disabled?: boolean;
  loading?: boolean;
  hoverClass?: string;
  hoverStartTime?: number;
  hoverStayTime?: number;
  hoverStopPropagation?: boolean;
  formType?: 'submit' | 'reset';
  openType?: 'share' | 'getPhoneNumber';
  onClick?: (e: any) => void;
  onGetphonenumber?: (e: any) => void;
}

const Button = factory<ButtonProps>('button');

export default Button;
