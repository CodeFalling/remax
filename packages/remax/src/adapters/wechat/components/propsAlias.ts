import plainStyle from '../../../utils/plainStyle';

const alias: { [prop: string]: string } = {
  activeColor: 'activeColor',
  backgroundColor: 'backgroundColor',
  onClick: 'bindtap',
  catchClick: 'catchtap',
  enable3D: 'enable-3D',
  hTouchMove: 'htouchmove',
  vTouchMove: 'vtouchmove',
};

function getAlias(prop: string) {
  const aliasProp = alias[prop];

  if (aliasProp) {
    return aliasProp;
  }

  if (prop.endsWith('className')) {
    return prop.replace('className', 'class');
  }

  if (prop.startsWith('on') || prop.startsWith('catch')) {
    return prop.toLowerCase().replace('on', 'bind');
  }

  return prop;
}

function getValue(prop: string, value: any): any {
  if (prop.endsWith('style') && prop !== 'layer-style') {
    return plainStyle(value);
  }

  return value;
}

export interface GenericProps {
  [key: string]: any;
}

export default function propsAlias<T>(props: GenericProps) {
  const aliasProps: GenericProps = {};

  Object.keys(props).forEach(prop => {
    aliasProps[getAlias(prop)] = getValue(prop, props[prop]);
  });

  return aliasProps;
}
