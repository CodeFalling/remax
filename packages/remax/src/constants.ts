export const REMAX_ROOT_BACKUP = '$$REMAX_ROOT_BACKUP';
export const TYPE_TEXT = 'plain-text';
