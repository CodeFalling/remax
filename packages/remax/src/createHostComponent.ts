import * as React from 'react';
import { TYPE_TEXT } from './constants';

export const hostComponents = new Set();

export function isHostComponent(type: string) {
  if (type === TYPE_TEXT) {
    return true;
  }

  return hostComponents.has(type);
}

export default function createHostComponent<P = any>(
  name: string,
  component?: React.ComponentType<P>
) {
  hostComponents.add(name);

  if (component) {
    return component;
  }

  const Component: React.FC<P> = (props, ref: any) => {
    const { children = [] } = props;
    return React.createElement(name, { ...props, ref }, children);
  };
  return React.forwardRef<{}, React.PropsWithChildren<P>>(Component);
}
