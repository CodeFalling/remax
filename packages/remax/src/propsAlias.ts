import { TYPE_TEXT } from './constants';
import kebabCase from 'lodash.kebabcase';
import Platform from './Platform';
import plainStyle from './utils/plainStyle';
import { isHostComponent as isWechatHostComponent } from './adapters/wechat/components/factory';
import { isHostComponent as isAlipayHostComponent } from './adapters/alipay/components/factory';
import { isHostComponent as isToutiaoHostComponent } from './adapters/toutiao/components/factory';

export function isHostComponent(name: string) {
  if (name === TYPE_TEXT) {
    return true;
  }

  switch (Platform.current) {
    case 'wechat':
      return isWechatHostComponent(name);
    case 'alipay':
      return isAlipayHostComponent(name);
    case 'toutiao':
      return isToutiaoHostComponent(name);
  }

  return false;
}

function functionPropAlias(prop: string, platform?: string) {
  prop = prop.replace(/Click$/, 'Tap');

  if (platform === 'alipay') {
    return prop;
  }

  prop = prop.toLowerCase().replace('on', 'bind');

  return prop;
}

export function getAlias(prop: string, isNative = false, platform?: string) {
  prop = prop.replace('className', 'class').replace('ClassName', 'Class');

  if (isNative) {
    return prop;
  }

  if (prop.startsWith('on') || prop.startsWith('catch')) {
    return functionPropAlias(prop, platform);
  }

  return kebabCase(prop);
}

function getValue(prop: string, value: any): any {
  if (prop.toLowerCase().endsWith('style') && prop !== 'layerStyle') {
    return plainStyle(value);
  }

  return value;
}

export interface GenericProps {
  [key: string]: any;
}

export default function propsAlias(
  props: GenericProps,
  isNative = false,
  platform = Platform.current
) {
  if (!props) {
    return props;
  }

  const aliasProps: GenericProps = {};

  Object.keys(props).forEach(prop => {
    aliasProps[getAlias(prop, isNative, platform)] = getValue(
      prop,
      props[prop]
    );
  });

  return aliasProps;
}
