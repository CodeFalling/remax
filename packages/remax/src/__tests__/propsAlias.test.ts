import { hostComponents } from 'remax.macro';
import propsAlias, { getAlias } from '../propsAlias';
import { isHostComponent } from '../createHostComponent';
import { TYPE_TEXT } from '../constants';

describe('props alias', () => {
  it('transform className prop correctly', () => {
    expect(getAlias('className')).toBe('class');
    expect(getAlias('placeholderClassName')).toBe('placeholder-class');
    expect(getAlias('className', true)).toBe('class');
    expect(getAlias('placeholderClassName', true)).toBe('placeholderClass');

    expect(
      propsAlias({
        className: 'class-name',
        placeholderClassName: 'placeholder-class-name',
      })
    ).toEqual({
      class: 'class-name',
      'placeholder-class': 'placeholder-class-name',
    });
    expect(
      propsAlias(
        {
          className: 'class-name',
          placeholderClassName: 'placeholder-class-name',
        },
        true
      )
    ).toEqual({
      class: 'class-name',
      placeholderClass: 'placeholder-class-name',
    });
  });

  it('transform function prop correctly', () => {
    expect(getAlias('onTouchStart')).toBe('bindtouchstart');
    expect(getAlias('onClick')).toBe('bindtap');
    expect(getAlias('catchClick')).toBe('catchtap');
    expect(getAlias('onLongClick')).toBe('bindlongtap');
    expect(getAlias('onClick', true)).toBe('onClick');

    expect(getAlias('onTouchStart', false, 'alipay')).toBe('onTouchStart');
    expect(getAlias('onClick', false, 'alipay')).toBe('onTap');
    expect(getAlias('catchClick', false, 'alipay')).toBe('catchTap');
    expect(getAlias('onLongClick', false, 'alipay')).toBe('onLongTap');
    expect(getAlias('onClick', true, 'alipay')).toBe('onClick');
  });

  it('transform style prop correctly', () => {
    expect(
      propsAlias({
        style: {
          color: '#ffffff',
          height: '2px',
        },
      })
    ).toMatchSnapshot();
  });

  it('transform props by component type correctly', () => {
    expect(getAlias('prop')).toBe('prop');

    hostComponents.set('foo', {
      props: ['bar', 'camelCase'],
      alias: {
        camelCase: 'kebab-case',
      },
    });

    expect(
      propsAlias(
        {
          bar: 'bar',
          camelCase: 'value',
        },
        false,
        undefined,
        undefined,
        'foo'
      )
    ).toEqual({
      bar: 'bar',
      'kebab-case': 'value',
    });
  });

  describe('check host components', () => {
    afterEach(() => {
      process.env['REMAX_PLATFORM'] = undefined;
    });

    it('wechat', () => {
      process.env['REMAX_PLATFORM'] = 'wechat';
      require('../adapters/wechat/components/Button');
      expect(isHostComponent(TYPE_TEXT)).toBeTruthy();
      expect(isHostComponent('button')).toBeTruthy();
      expect(isHostComponent('foo')).toBeFalsy();
    });

    it('alipay', () => {
      process.env['REMAX_PLATFORM'] = 'alipay';
      require('../adapters/alipay/components/Button');
      expect(isHostComponent(TYPE_TEXT)).toBeTruthy();
      expect(isHostComponent('button')).toBeTruthy();
      expect(isHostComponent('foo')).toBeFalsy();
    });

    it('toutiao', () => {
      process.env['REMAX_PLATFORM'] = 'toutiao';
      require('../adapters/toutiao/components/Button');
      expect(isHostComponent(TYPE_TEXT)).toBeTruthy();
      expect(isHostComponent('button')).toBeTruthy();
      expect(isHostComponent('foo')).toBeFalsy();
    });
  });
});
