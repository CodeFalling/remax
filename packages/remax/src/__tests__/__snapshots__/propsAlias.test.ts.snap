// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props alias transform style prop correctly 1`] = `
Object {
  "style": "color:#ffffff;height:2rpx;",
}
`;
